pkg/remoteimage/pull.go

```diff
@@ -31,23 +35,76 @@ type puller struct {
 
 // Returns the compressed size of the image that was pulled in bytes
 // see https://github.com/containerd/containerd/issues/9261
-func (p puller) ImageSize(ctx context.Context) int {
+func (p puller) ImageSize(ctx context.Context) (size int, err error) {
+	defer func() {
+		if err != nil {
+			klog.Errorf(err.Error())
+			metrics.OperationErrorsCount.WithLabelValues("size-error").Inc()
+		}
+	}()
 	imageSpec := &cri.ImageSpec{Image: p.image.String()}
-	imageStatusResponse, _ := p.imageSvc.ImageStatus(ctx, &cri.ImageStatusRequest{
+	if imageStatusResponse, err := p.imageSvc.ImageStatus(ctx, &cri.ImageStatusRequest{
 		Image: imageSpec,
-	})
-	return int(imageStatusResponse.Image.Size_)
+	}); err != nil {
+		size = 0
+		err = fmt.Errorf("remoteimage.ImageSize(): call returned an error: %s", err.Error())
+		return size, err
+	} else if imageStatusResponse == nil {
+		size = 0
+		err = fmt.Errorf("remoteimage.ImageSize(): imageStatusResponse is nil")
+		return size, err
+	} else if imageStatusResponse.Image == nil {
+		size = 0
+		err = fmt.Errorf("remoteimage.ImageSize(): imageStatusResponse.Image is nil")
+		return size, err
+	} else {
+		size = imageStatusResponse.Image.Size()
+		err = nil
+		return size, err
+	}
 }
 
 func (p puller) Pull(ctx context.Context) (err error) {
+	startTime := time.Now()
+	defer func() { // must capture final value of "err"
+		elapsed := time.Since(startTime).Seconds()
+		// pull time metrics and logs
+		klog.Infof("remoteimage.Pull(): pulled %s in %d milliseconds", p.image.String(), int(1000*elapsed))
+		metrics.ImagePullTimeHist.WithLabelValues(metrics.BoolToString(err != nil)).Observe(elapsed)
+		metrics.ImagePullTime.WithLabelValues(p.image.String(), metrics.BoolToString(err != nil)).Set(elapsed)
+		if err != nil {
+			metrics.OperationErrorsCount.WithLabelValues("pull-error").Inc()
+		}
+		go func() {
+			//TODO: this is a hack to ensure data is cleared in a reasonable time frame (after scrape) and does not build up.
```

Comment:
this hack is mentioned above. spawns go routine to sleep. could instead add to a channel with a timestamp and an anonymous func. would have to guard against deadlock scenarios.